content/courses/advanced-ada/parts/data_types/numerics.rst

```diff
@@ -1329,22 +1329,457 @@ input parameter :ada:`V` (of :ada:`One_To_Ten'Base` type) in the runtime output
+
+The compilation of this package fails because the :ada:`Custom_Int'First` is
+below :ada:`System.Min_Int` and :ada:`Custom_Int'Last` is above
+:ada:`System.Max_Int`.
+
+
+Range of base type
+^^^^^^^^^^^^^^^^^^
+
+As we've said before, a custom type definition is derived from the root
+integer type. The range of its :ref:`base type <Adv_Ada_Base_Types>`, however,
+is *not* derived from the root integer type, but rather determined by the range
+of the type specification. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Numerics.Discrete_Numeric_Types.Custom_Int_Base_Range
+
+    with System;
+
+    package Custom_Integer_Types is
+
+       type Custom_Int is
+         range 1 .. 10;
+
+    end Custom_Integer_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Custom_Integer_Types;
+    use  Custom_Integer_Types;
+
+    procedure Show_Custom_Integer_Types is
+    begin
+       Put_Line ("Custom_Int'Size  :"
+                 & Custom_Int'Size'Image
+                 & " bits");
+       Put_Line ("Custom_Int'First :"
+                 & Custom_Int'First'Image);
+       Put_Line ("Custom_Int'Last  :"
+                 & Custom_Int'Last'Image);
+
+       Put_Line ("Custom_Int'Base'Size  :"
+                 & Custom_Int'Base'Size'Image);
+       Put_Line ("Custom_Int'Base'First :"
+                 & Custom_Int'Base'First'Image);
+       Put_Line ("Custom_Int'Base'Last  :"
+                 & Custom_Int'Base'Last'Image);
+    end Show_Custom_Integer_Types;
+
+On a typical desktop PC, you might see that the range of :ada:`Custom_Int'Base`
+is :ada:`-128 .. 127`, while the
+:ref:`system max. and min. values <Adv_Ada_System_Min_Max_Int>` we've seen
+before had a much wider range.
+
+As a reminder, the range of the base type might be wider than the range of the
+custom integer type we're defining. (We mentioned this earlier on when
+discussing :ref:`base types <Adv_Ada_Base_Types>`.)
+
+
+.. _Adv_Ada_Modular_Types:
+
+Modular Types
+~~~~~~~~~~~~~
+
+As we've mentioned in the :ref:`Introduction to Ada <Intro_Ada_Unsigned_Types>`
+course, modular types are the Ada version of *unsigned* integer types. We
+declare a modular type by specifying its modulo |mdash| by using the
+:ada:`mod` keyword:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Data_Types.Numerics.Discrete_Numeric_Types.Modular_Types
+
+    package Modular_Types is
+
+       type Modular is mod 2**32;
+
+    end Modular_Types;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    with Modular_Types;
+    use  Modular_Types;
+
+    procedure Show_Modular_Types is
+    begin
+       Put_Line ("Modular'Size  :"
+                 & Modular'Size'Image
+                 & " bits");
+       Put_Line ("Modular'First :"
+                 & Modular'First'Image);
+       Put_Line ("Modular'Last  :"
+                 & Modular'Last'Image);
+    end Show_Modular_Types;
+
+This example declares the 32-bit modular type :ada:`Modular`.
+
```

Comment:
Probably should emphasize that the specified modulus need not be a power of 2. I have found that C programmers automatically assume that restriction.